src/Utilities/System/Prefetch.hpp

```diff
@@ -0,0 +1,38 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <iosfwd>
+
+namespace sys {
+/// The cache location to prefetch data to.
+enum class PrefetchTo : int {
+  /// Prefetch into L1 data cache
+  L1Cache = 3,
+  /// Prefetch into L2 data cache
+  L2Cache = 2,
+  /// Prefetch into L3 data cache
+  L3Cache = 1,
+  /// Non-temporal is an element that is unlikely to be re-used. E.g., read
+  /// once, written but never read.
+  NonTemporal = 0,
+  /// Prefetch to L1 data cache for writing
+  WriteL1Cache = 7,
+  /// Prefetch to L2 data cache for writing
+  WriteL2Cache = 6
+};
+
+/// \brief Prefetch data into a specific level of data cache.
+template <PrefetchTo CacheLocation>
+#if defined(__GNUC__)
+__attribute__((always_inline)) inline
+#endif
+void prefetch(const void* address_to_prefetch) {
```

Comment:
No, because then you could have invalid states that I don't know what they would do... The list isn't a tensor product of the cache level and write-only.